ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java

```diff
@@ -233,101 +187,61 @@ public int getShiftedPos(int pos) {
-            this.endOffset = endOffset;
-
-            SequenceElement.ComparisonItem fromInclusive = new SequenceElement.ComparisonItem(startOffset);
-            SortedSet<SequenceElement> subMap = colorings.tailSet(fromInclusive);
-            iterator = subMap.iterator();
-        }
-
-        private SequenceElement fetchElementFromIterator(boolean updateNextElementStartOffset) {
-            int seStartOffset;
-            SequenceElement se;
-            if (iterator != null && iterator.hasNext()) {
-                se = iterator.next();
-                seStartOffset = se.range.getStart();
-                if (LOG.isLoggable(Level.FINE)) {
-                    LOG.log(Level.FINE, "Fetched highlight <{0},{1}>\n", // NOI18N
-                            new Object[]{seStartOffset, se.range.getEnd()});
-                }
-                if (seStartOffset >= endOffset) {
-                    se = null;
-                    seStartOffset = Integer.MAX_VALUE;
-                    iterator = null;
-                }
-            } else {
-                se = null;
-                seStartOffset = Integer.MAX_VALUE;
-                iterator = null;
-            }
-            if (updateNextElementStartOffset) {
-                nextElementStartOffset = seStartOffset;
-            }
-            return se;
+        GsfHighlightSequence(Iterator<SequenceElement> it) {
+            iterator = it;
         }
 
         @Override
         public boolean moveNext() {
-            if (nextElement != null) {
-                element = nextElement;
-                nextElement = fetchElementFromIterator(true);
-            } else {
-                if ((element = fetchElementFromIterator(false)) != null) {
-                    nextElement = fetchElementFromIterator(true);
-                }
-            }
-            return (element != null);
+            element = iterator.hasNext() ? iterator.next() : null;
+            return element != null;
         }
 
         @Override
         public int getStartOffset() {
-            return (element != null)
-                    ? layer.getShiftedPos(element.range.getStart())
-                    : Integer.MAX_VALUE;
+            return getShiftedPos(element.range().getStart());
         }
 
         @Override
         public int getEndOffset() {
-            return (element != null)
-                    ? Math.min(layer.getShiftedPos(element.range.getEnd()), nextElementStartOffset)
-                    : Integer.MAX_VALUE;
+            return getShiftedPos(element.range().getEnd());
         }
 
         @Override
         public AttributeSet getAttributes() {
-            return (element != null)
-                    ? layer.getColoring(element.coloring, element.language)
-                    : SimpleAttributeSet.EMPTY;
+            return getColoring(element.coloring(), element.language());
         }
     }
+
+    private static <T> int firstSequenceElement(List<SequenceElement> l, int offset) {
```

Comment:
A comment what this method should do would be nice. I have a feeling, but knowing would be nice.